[SUREFIRE-756] Allow ability to capture executed random runOrder for re-play purposes

winglam opened the pull request:
The changes in this pull request addresses issue SUREFIRE-756. Namely, the changes include

- Output of the random seed used to generate a particular random test order when -Dsurefire.runOrder=random or -Dfailsafe.runOrder=random is set
- Ability to replay a previously observed random test order by setting -Dsurefire.seed and -Dfailsafe.seed to the seed that observed the random test order
- Tests to ensure that the setting of the _same_ random seeds do create the _same_ test orders and _different_ random seeds do create _different_ test orders. Note that the inherent randomness of the orders does mean that the tests can be flaky (nondeterministically pass or fail without changes to the code). The current tests have a rate of 0.4% ```(1/3)^5``` of failing. Increasing the number of tests (```3```) or the number of times to loop (```5```) would decrease the odds of the tests failing.

Tibor17 commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java:
```diff
@@ -310,6 +310,22 @@
+     * @since 3.0.0-M6
+     */
+    @Parameter( property = "failsafe.seed" )
+    private Long randomSeed;
```
Did you take over the changes from an old PR?
We spoke about this parameter that we already have the config parameter named `runOrder`. And so that it should contain the seed too, let's suppose `<runOrder>random1234567890</runOrder>`.
The reason is that we have too many config parameters and this one is not urgent due to we are able to specify the seed in one business parameter `runOrder`.

winglam replied:
These changes were not taken from an old PR, but I would be happy to take a look and compare the changes if you can point me to the PR.

As for not adding failsafe.seed or surefire.seed, I can change it so that it just parses the numbers after random as the seed instead. Do let me know if that's what is desired here.

Tibor17 replied:
> Do let me know if that's what is desired here.

Yes this is what I mean. Basically, it is easy to parse the number. You can use regex or the string spliterator.

rmannibucau replied:
-1, it is two separate params and must be overridable with a system prop easily and *independently* so it is good like that.
You can move it to other parameters or even a map of runOrderParameters to avoid more params but having to merge the steing is bad for end users and not that consistent with the config IMHO.

Tibor17 replied:
@rmannibucau This was we would have too many config params. Having it in one means that it is about the same business. Here you are setting only run order and so that you say "random" is the same business value with the "seed".
Alternatively it can be done another way with the extensions where this parameter may become an interface. It means that the parameter "runOrder" would become an element what is called complexType in the XML and it may have specific nested elements for the concrete implementation of the interface, and the "seed" would be only for the "random".

rmannibucau replied:
@Tibor17 fact it creates less parameters is irrelevant - in particular today we have a tons of parameters. Fact it makes it harder to use, integrate and is more error prone for end users is what counts at the end and here having a dedicated variable for the replay is highly relevant IMHO. Also note that having the seed is not always enough to replay a test suite (in particular in random mode) so you can just need to capture the test order and replay it explicitly (a runOrder=fromFile and runOrderFile=/path would be more relevant there).

winglam replied:
> capture the test order and replay it explicitly (a runOrder=fromFile and runOrderFile=/path would be more relevant there)
I actually have some changes (in a separate local branch) that would accomplish this as well and plan to submit it after this PR is merged. I could still use some help/advice on how to capture the test order (with or without running tests). The goal of my changes would be to empower runOrder to be more customizable and deterministic with seeds, passing in a file, controlling not just test class order but also test method order, etc. 

I have the changes for merging `failsafe.seed` and `surefire.seed` with the `runOrder` parameter ready, but will leave those unpushed for now as I am fine with merging them or keep them separate (as they are now or moving it, etc.).

Tibor17 replied:
@winglam 
Pls refactor the name `randomSeed` and the property name `failsafe.seed`, and for surefire too.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java:
```diff
@@ -3051,6 +3056,17 @@ private void warnIfIllegalTempDir() throws MojoFailureException
         }
     }
 
+    private void printDefaultSeedIfNecessary()
```
This abstract MOJO class is used in the Mockito tests. Can you cover all your changes and extend the existing tests? It would be cool to have tests maybe for larger methods too.

winglam replied:
I can certainly add a test or two that covers this method. If there are other methods that I should cover or if there are particular tests that I should extend, do let me know.

Tibor17 replied:
These are the test classes, `AbstractSurefireMojoJava7PlusTest` and `AbstractSurefireMojoTest`, where we add the tests for `AbstractSurefireMojo`.

Tibor17 commented on maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/booterclient/BooterDeserializerProviderConfigurationTest.java:
```diff
@@ -276,7 +276,7 @@ private ProviderConfiguration getTestProviderConfiguration( DirectoryScannerPara
                              new TestListResolver( USER_REQUESTED_TEST + "#aUserRequestedTestMethod" ),
                     RERUN_FAILING_TEST_COUNT );
-        RunOrderParameters runOrderParameters = new RunOrderParameters( RunOrder.DEFAULT, null );
+        RunOrderParameters runOrderParameters = new RunOrderParameters( RunOrder.DEFAULT, null, null );
```
It would be worth to test the non-null value.

Tibor17 commented on maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/booterclient/BooterDeserializerStartupConfigurationTest.java:
```diff
@@ -197,7 +197,7 @@ private ProviderConfiguration getProviderConfiguration()
                              new TestListResolver( "aUserRequestedTest#aUserRequestedTestMethod" ) );
 
-        RunOrderParameters runOrderParameters = new RunOrderParameters( RunOrder.DEFAULT, null );
+        RunOrderParameters runOrderParameters = new RunOrderParameters( RunOrder.DEFAULT, null, null );
```
detto here

Tibor17 commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/SurefireReflector.java:
```diff
@@ -178,10 +178,11 @@ private Object createRunOrderParameters( RunOrderParameters runOrderParameters )
         Constructor<?> constructor = getConstructor( this.runOrderParameters, arguments );
         File runStatisticsFile = runOrderParameters.getRunStatisticsFile();
-        return newInstance( constructor, RunOrder.asString( runOrderParameters.getRunOrder() ), runStatisticsFile );
+        return newInstance( constructor, RunOrder.asString( runOrderParameters.getRunOrder() ), runStatisticsFile,
```
This reflector should be tested in `SurefireReflectorTest`.

winglam commented:
@Tibor17 just a reminder that since we last spoke, I've proofread my changes again, did the refactoring you asked, cleaned up some unnecessary changes, and added a test for the SurefireReflector. If there's anything else needed, let me know and I'll try to get to it soon

Tibor17 commented:
I will have a look in few hours. T

Dňa št 17. 9. 2020, 0:30 winglam <notifications@github.com> napísal(a):

> @Tibor17 <https://github.com/Tibor17> just a reminder that since we last
> spoke, I've proofread my changes again, did the refactoring you asked,
> cleaned up some unnecessary changes, and added a test for the
> SurefireReflector. If there's anything else needed, let me know and I'll
> try to get to it soon
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/maven-surefire/pull/309#issuecomment-693699949>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH7ER3EGY55ND5PWQLF2B3SGE4BTANCNFSM4PQ6K2MA>
> .
>

Tibor17 commented:
@winglam 
First of all the build must be stable, otherwise adding more and more commits would be a disaster.
The build of master is in progress, see https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-surefire/job/master/
